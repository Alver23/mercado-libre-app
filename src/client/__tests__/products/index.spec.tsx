// Dependencies
import { shallow, ShallowWrapper } from 'enzyme';
import { GetServerSidePropsContext, GetServerSidePropsResult } from 'next';

// Services
import ProductService from '@alversoft/client/services/products';

// Under test file
import ProductList, { getServerSideProps } from '@alversoft/pages/products';

// Mocks
import mocks from './mocks.json';

describe('<ProductList />', () => {
  let component: ShallowWrapper;
  const props = {
    ...mocks,
  };
  beforeEach(() => {
    component = shallow(<ProductList {...props} />);
  });

  it('should save a snapshot of the component', () => {
    expect(component).toMatchSnapshot();
  });

  describe('getServerSideProps method', () => {
    const searchText = 'apple ipod';
    const context: Partial<GetServerSidePropsContext> = {
      query: {
        q: searchText,
      },
    };

    afterEach(() => {
      expect(ProductService.getProducts).toHaveBeenCalledWith(context.query.q);
      expect(ProductService.getProducts).toHaveBeenCalledTimes(1);
    });

    it('should get the props successfully', async () => {
      jest.spyOn(ProductService, 'getProducts').mockResolvedValue({ ...mocks.data });
      const response: GetServerSidePropsResult<unknown> = await getServerSideProps(
        context as GetServerSidePropsContext
      );
      expect(response).toHaveProperty('props.data', mocks.data);
      expect(response).toHaveProperty('props.searchText', searchText);
    });

    it('should get an error when service fails', async () => {
      const exceptionMocked = new Error('fake error');
      jest.spyOn(ProductService, 'getProducts').mockRejectedValue(exceptionMocked);
      const response: GetServerSidePropsResult<unknown> = await getServerSideProps(
        context as GetServerSidePropsContext
      );
      expect(response).toHaveProperty('notFound', true);
    });
  });
});
