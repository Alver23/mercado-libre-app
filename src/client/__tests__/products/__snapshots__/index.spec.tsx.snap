// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductList /> should save a snapshot of the component 1`] = `
<Layout
  meta={
    <Meta
      title="fake search Text"
    />
  }
>
  <div
    className="container mx-auto flex justify-center items-center"
  >
    <div
      className="w-11/12 lg:w-10/12"
    >
      <div>
        <Breadcrumb
          items={
            Array [
              "Fake text 1",
              "fake test 2",
            ]
          }
        />
      </div>
      <section
        className="bg-white px-6 rounded"
      >
        <ProductList
          items={
            Array [
              Object {
                "condition": "new",
                "freeShipping": true,
                "id": "MLA853662269",
                "picture": "http://http2.mlstatic.com/D_922186-MLA41184540762_032020-I.jpg",
                "price": Object {
                  "amount": 19999,
                  "currency": "ARS",
                  "decimals": 2,
                },
                "title": "Auriculares In-ear Bose Soundsport In-ear Headphones Charcoal",
              },
            ]
          }
        />
      </section>
    </div>
  </div>
</Layout>
`;
